src/controller/CHIPDeviceControllerFactory.h

```diff
@@ -138,6 +138,8 @@ class DeviceControllerFactory
     //
     void ReleaseSystemState() { mSystemState->Release(); }
 
+    const DeviceControllerSystemState * GetSystemState() { return mSystemState; }
```

Comment:
So if you call this and register a handler, but then the state gets torn down (because the number of live CHIPDeviceController instances goes to 0), what should happen?

Put another way, should this API be on the controller factory, or on the controller?

Or should we actually have an API where unsolicited message handlers can be registered via the controller factory, and then set up on whatever exchange manager we have when we have it?

Reply:
> Put another way, should this API be on the controller factory, or on the controller?

One valid use case to have a getter in the factory is when you aren't using controllers the factory is vending, but want the factory to setup the singleton-like objects a custom controller would need, and then want to access these through mSystemState.

As to how this needs to work, I think the consumer of this just needs to retain / release like the sdk's controllers, or needs to make sure the factory itself isn't destroyed while it is in use.

As to what we could do instead, `chip::Platform::SharedPtr<T>` is a potential tool.

Reply:
> Put another way, should this API be on the controller factory, or on the controller?

I think it's an odd pattern to go to a in-direct consumer's view of these singleton objects, or be required to vend a controller to interact with these singletons.

> Or should we actually have an API where unsolicited message handlers can be registered via the controller factory

That feels like an anti-pattern, to 'forward' selective functions from the underlying object through. That doesn't scale.

> but then the state gets torn down

Either you call `RetainSystemState` to prevent that (accompanied by a matching ReleaseSystemState), or you sequence your shutdown to ensure that any objects that utilize this method of registering unsolicited handlers with the exchange manager is shut down first before you shutdown your controller.

I think that's best left to the application to figure out.

Reply:
I'd be good with adding verbiage around this as comments above.

Reply:
I'll note that today, we provide getters to get at `ExchangeManager` from a myriad of objects in the SDK today, so clearly, there is a need to interact with it more directly.

Reply:
Jerry and Michael, I see you've both approved the PR. Can we mark this conversation resolved?

Reply:
My vote is yes.  However, @bzbarsky-apple raised the question.

Reply:
Ah, good point about `RetainSystemState`.  And thank you for adding the comments about it!